README.md
# Simultaneous Combat System

![Version](https://img.shields.io/github/v/tag/arcanistzed/scs?label=Version&style=flat-square&color=2577a1) ![Latest Release Download Count](https://img.shields.io/github/downloads/arcanistzed/scs/latest/module.zip?label=Downloads&style=flat-square&color=9b43a8) ![Supported Foundry Versions](https://img.shields.io/endpoint?url=https://foundryshields.com/version?url=https://raw.githubusercontent.com/arcanistzed/scs/main/module.json&style=flat-square&color=ff6400) [![Discord Server](https://img.shields.io/badge/-Discord-%232c2f33?style=flat-square&logo=discord)](https://discord.gg/AAkZWWqVav) [![Patreon](https://img.shields.io/badge/-Patreon-%23141518?style=flat-square&logo=patreon)](https://www.patreon.com/bePatron?u=15896855)

A beautiful mostly system-agnostic and completely customizable implementation of a Simultaneous Combat System. See the [Reddit post](https://redd.it/p11h35) from the author.

Have you ever tired of waiting for the turns of your friends to be over? Are you fed up with your players taking so long in *every* combat encounter? Do you feel like changing things up in your game and trying something new?

![gradient](https://user-images.githubusercontent.com/82790112/123046244-ddeb3a80-d3c9-11eb-98db-da2f4a6abd68.gif)

## Installation

In the setup screen, use the URL `https://github.com/arcanistzed/scs/releases/latest/download/module.json` to install the module.

## Usage

Please follow the built-in tutorial (built with IntroJS), for instructions. You can always relaunch the tutorial from within the settings. If you have any questions, feel free to contact me.

## Support

Please consider supporting me on [my Patreon](https://patreon.com/arcanistzed) if you like my work. You can see a list of all my projects on [my website](https://arcanist.me).

## Features

### Action Locking

In supported systems, SCS will detect whether the current phase has been registered with the module and will restrict what can be done on your turn accordingly.

If you would like to add support for another system or more phases, please make a PR or let me know on my Discord (link is below) and I will happily take suggestions.

Currently only DnD5e is supported with the default phases as described in the reddit post linked above.

When using BetterRolls5e or MidiQoL, you will get this warning from libWrapper which is safe to ignore:

```txt
libWrapper: Potential conflict detected between module 'scs' and module 'betterrolls5e'.
Module 'scs' did not chain the wrapper for 'CONFIG.Item.documentClass.prototype.roll'.
```

### Phase and Round display

The module will display the current phase and round in an interactive and draggable app. You can use the arrows to change phase (`<` or `>`) and change round (`<<` or `>>`), or you can simply click on the phase you would like to switch to (not recommended if using phase/cycle limits).

### Integration with the Core Combat Tracker

This module updates the Combat Tracker as you use it for better compatibility, but hides the buttons for changing turns. This module also prevents any turns to be assigned, so that you don't see any highlighting for such in the Combat Tracker.
SCS prompts to end combat whenever the round is equal to zero and will start a new combat if there is none when loading. The module will end the combat when all the combatants are removed.
There is also a setting to hide the default Combat Tracker.

### Hides itself when there is no Combat

When enabled, once the combat has ended or all of the combatants have been removed, the app will hide itself. It will show itself again when a new combat is created.
You can use this feature without the default Combat Tracker by simply adding and removing combatants in the Token HUD when combat is beginning or is over.

### Attack Roll Display

In order to ease determining the attack order (for example, in the default attack phase), the module shows you the last attack roll either next to the token or in the Combat Tracker. This currently only works in DnD5e and PF2e.

### Keybindings

You can switch the phase, current round, or toggle the app's visibility with keybindings.

### Integration with other modules

#### [SmallTime](https://foundryvtt.com/packages/smalltime)

The app moves outside of SmallTime's way and locks into place right above it.

#### [Simple Calendar](https://foundryvtt.com/packages/foundryvtt-simple-calendar) and [About Time](https://foundryvtt.com/packages/about-time)

Works alongside the realtime clock by manipulating the Combat Tracker in the background, therefore the clock will be paused when the round is not zero and the game time will be changed when the round is changed.

#### [Argon Combat HUD](https://foundryvtt.com/packages/enhancedcombathud)

This module hides Argon's "End Turn" button.

#### [Better Rolls 5e](https://foundryvtt.com/packages/betterrolls5e)

The Attack Display parses and handles Better Rolls' chat messages to allow it to display attack rolls made with the module.
Please note that SCS may have difficulty determining the correct token to add the display to if you use Better Rolls to roll for linked tokens. To get the right result, you can select the token before rolling for attack. If that doesn't work, make sure you only have one token from that Actor on the map.

### Configure Colors

A configuration menu that allows you to manually configure all colors used. The numbers in the fields represent the hue value (from 0 to 360) of the top and bottom of each linear gradient. You can click the dice buttons to generate a random value for that field.

![configuration](https://i.imgur.com/7LFwm7R.png)

### Limit Phases

This setting will automatically switch the round at the end of the cycle of all phases. You can also do this with Maximum Cycle set to 1 and Limit Cycles enabled.

### Limit Cycles & Maximum Cycle

This setting will put a limit on how many cycles can take place before switching to the next round. For example, if Maximum Cycle is set to 3, after each phase has been completed three times, it will be the next round. It is recommended using the arrows to navigate between phases when using this.

## API

### Methods

SCS has a small API with a few methods that can be called by others. They are accessible under `game.modules.get("scs")?.api`.

#### `changeVisibility(state)`

This will change the visibility of the SCS Application. You cen either set `state` to `true` to display the App or to `false` to hide it.

#### `defaultTracker(hide = true)`

This will elegantly hide the default combat tracker from the sidebar. An optional Boolean parameter can be used to show the tracker again, if hidden.

#### `startTutorial()`

This will show the IntroJS tutorial tour once.

#### `stopTutorial(close = false)`

This will stop the tour from showing every time the page is loaded unless the user re-enabled the tutorial from within the module settings.
An optional Boolean parameter can be used to also immediately close the tutorial, but this defaults to `false`.

#### Deprecated: `changeRound(delta)`

~~An asynchronous method that changes the SCS's round by a given delta. Note that this will also change the Core round. Use a positive number for the delta to move the round forward and a negative number to go to previous rounds.~~

Use `game.combat.nextRound()` or `game.combat.previousRound()` instead.

#### `changePhase(delta)`

An asynchronous method that changes the SCS's phase by a given delta. Use a positive number for the delta to move the phase forward and a negative number to go to previous phase.

### Hooks

This module has two hooks which are only called on the GM's client:

#### Deprecated: `scsRoundChanged`

~~This hook is called whenever the SCS round changes. It has three arguments which are (in order): the current phase, the previous phase, and the delta.~~

Use the `updateCombat` hook and check the `change.round` value instead.

#### `scsPhaseChanged`

This hook is called whenever the SCS phase changes. It has three arguments which are (in order): the current phase, the previous phase, and the delta.

## Bugs

You can submit bugs via [Github Issues](https://github.com/arcanistzed/scs/issues/new/choose) or on [my Discord server](https://discord.gg/AAkZWWqVav).

## Localization

Please note that the `Phases` must be short enough to fit their boxes, so you shouldn't put anything over approximately eight characters.

## Contact me

Come hang out on [my Discord server](https://discord.gg/AAkZWWqVav) or [click here to send me an email](mailto:arcanistzed@gmail.com?subject=Simultaneous%20Combat%20System%20module).

## License

Copyright © 2021 arcanist

This package is under an [MIT license](LICENSE) and the [Foundry Virtual Tabletop Limited License Agreement for module development](https://foundryvtt.com/article/license/).

See the [Notice](./NOTICE.md) for attribution details.


CHANGELOG.md
# Changelog

## 2.14.0 - 5 Jul 2022

### Added

Add configuration open for turn removal feature (enabled by default) [#14](https://github.com/arcanistzed/scs/issues/14)

### Fixed

CSS padding issue that caused arrow buttons to not be visible

## 2.13.0 - 4 Jul 2022

Removed the default keys for the keybinding for better compatibility (users can define their own)

### Added

* As part of the "display" feature, a flag is now set on actors with their last attack roll. This could be used in a custom [Combat Tracker Groups](https://github.com/arcanistzed/ctg) grouping mode, for example.
* Refactor CSS completely
  * New scroll snapping
* Configuration form to enable complete manual configuration of all colors

### Fixed

* Refactors to color generation and display
* Make phase and round switching more predictable
* Don't start combat if the Combat Tracker is not hidden (thanks to @Lethay)

## 2.12.0 - 13 Dec 2021

### Added

* Localized everything to make the "display" and "Action Locking" features work in other languages

### Fixed

* Removed separate round value since it was causing bugs with de-sync and making it so that you aren't able to advance the round at all. This separation was created because I originally wanted the SCS to work without any core combats active, but that no longer makes sense because they are now synchronized anyway

## 2.11.2 - 11 Dec 2021

### Fixed

* Error with display enabled when there is no flavor text in rolls
* Remove conflicting Keybindings

## 2.11.1 - 7 Dec 2021

### Fixed

* Use MidiQoL's hook
* Only attempt to add the Combat Tracker display if a combatant exists

## 2.11.0 - 7 Dec 2021

* IntroJS is now a soft dependency which you only need to install for the tutorial
* Inverted default Combat Tracker visibility setting

### Added

* Removed the concept of turns from the Combat Tracker:
  * Previous and Next turn buttons are hidden
  * No combatant is highlighted as the current combatant as there is no one taking their turn (requires v9d2 or later)
* Keybindings for switching phase and round as well as toggling app visibility (requires v9d2 or later)
* Major rework of the HUD feature which has been renamed to the more generic "Attack Display" and now includes an option to display the last roll in the Combat Tracker

### Fixed

* Always move above SmallTime no matter what it's height happens to be
* Attack Display works with MidiQoL
* Attack Display HUD is more consistently centered

## 2.10.0 - 19 Nov 2021

### Added

* Hide App when there is no Combat encounter
* End combat when all of the combatants have been removed

### Fixed

* Create a new combat if one doesn't exist if the round isn't zero
* Only allow GM users to change the phase or round with the API
* Bug fixes and improvements
* Code Quality improvements
* Fixed default settings in v9d2 which were messed up now that arrays can't be localized

## 2.9.0 - 24 Sept 2021

* *Action Locking:* the notification for an unrecognized phase is now GM-only
* *Combat Management:*
  * Added a prompt to end the current combat when the round becomes zero
  * Create a new Combat whenever the round becomes greater than zero
* *Color Generation:* player clients no longer reload after a delay
* Now always reloads the page after the phase names are changed in order to keep clients in sync
* Use v9 CSS variables for styling

## 2.8.0 - 16 Sept 2021 - Module Compatibility

* The HUD is now compatible with the [Better Rolls 5e](https://foundryvtt.com/packages/betterrolls5e) module (Fixes #5). See the README for details.
* Improved compatibility with [SmallTime](https://foundryvtt.com/packages/smalltime), to allow them to both position correctly together
* Some code cleanup and now there's no more jQuery!

## 2.7.2 - 6 Sept 2021

Set z-index above control toolbar

## 2.7.1 - 4 Sept 2021 - Window Closing

* Fix window closing by removing app from `ui.windows`
* Fix typos

## 2.7.0 - 30 Aug 2021

* Add `scsPhaseChanged` and `scsRoundChanged` hooks
* Use PIXI for the HUD to make it always show up. In addition, the HUD was redesigned. It now also only shows up during the Attack phase.

## 2.6.0 - 29 Aug 2021 - Attack Roll HUD

In order to ease determining the attack order (for example, in the default attack phase), the module shows you the last attack roll in the token's HUD. This currently only works in D&D 5e and PF 2e.

## 2.5.0 - 26 Aug 2021 - More API and Fixes

* Exposed round (`.changeRound(delta)`) and phase (`.changePhase(delta)`) switching in the public API
* Fix app border appearing incorrectly
* Updated [libWrapper](https://github.com/ruipin/fvtt-lib-wrapper) shim with dialog removed
* Migrated API to `game.modules.get("scs")?.api`
* Added optional parameter to `hideTracker` (which has been renamed to `defaultTracker`), allowing you to instead show the tracker if hidden
* Added optional parameter to `stopTutorial()` allowing you to also close the tutorial at the same time
* Fix a bug where the phase would be stuck in an invalid state after the names are changed
* Updated localisations

## 2.4.6 - 15 Aug 2021

* Fix action locking for Magic phase
* Removed [Argon's](https://foundryvtt.com/packages/enhancedcombathud) "End Turn" button

## 2.3.0 - 13 Aug 2021 - Cycles & Smooth scrolling

* Implemented cycles (which are one complete run through all phases) in order to allow for the round to end after three as per [SCSv2](https://redd.it/p11h35)
* Added animated smooth scrolling when there are too many phases to fit in the box

## 2.2.1 - 13 Aug 2021

Fix typo and fix not working on Firefox

## 2.2.0 - 12 Aug 2021 - Stacks


scripts/app.js
// Import libWrapper shim
import { libWrapper } from "./shim.js";

// Import public API
import api from "./api.js";

/**
 * Manages the SCS form app itself and it's functionality
 */
export default class scsApp extends FormApplication {
	/** Pixels by which to offset the app */
	static pinOffset = 50;

	/** The module's ID */
	static ID = "scs";

	/**
	 * An object containing the current phase names and colors
	 */
	static phases = {
		names: [],
		colors: [],
		count: undefined,
	};

	/** The current phase */
	static currentPhase = 1;

	/** The current cycle */
	static currentCycle = 1;

	/** A boolean representing whether SCS thinks there is currently a combat which is used for About Time integration
	 * If a new combatant is created, this will become true.
	 * If all combatants are removed, this will return to false.
	 */
	static inCombat = false;

	/** @override */
	static get defaultOptions() {
		const pinned = game.settings.get(scsApp.ID, "pinned");

		const playerApp = document.getElementById("players");
		const playerAppPos = playerApp.getBoundingClientRect();

		this.initialPosition = game.settings.get(scsApp.ID, "position");

		// The actual pin location is set elsewhere, but we need to insert something
		// manually here to feed it values for the initial render.
		if (pinned) {
			let rowsHeight = 46 * Math.min(Math.ceil(scsApp.phases.count / 3), 3); // Height of all rows of phases

			this.initialPosition.top = playerAppPos.top - scsApp.pinOffset - rowsHeight + 12;
			this.initialPosition.left = playerAppPos.left;
		}

		return mergeObject(super.defaultOptions, {
			template: "modules/scs/templates/app.hbs",
			id: "scsApp",
			title: "Simultaneous Combat System",
			classes: ["scs"],
			top: this.initialPosition.top,
			left: this.initialPosition.left,
			closeOnSubmit: false,
		});
	}

	/** @override */
	activateListeners([html]) {
		super.activateListeners(this.element);

		// Remove the app from `ui.windows` to not let it close with the escape key
		delete ui.windows[this.appId];

		// Make sure the app is pinned properly once loaded
		if (game.settings.get(scsApp.ID, "pinned")) scsApp.pinApp();

		// Make the app draggable
		const drag = new Draggable(this, this.element, document.querySelector("#scsApp .round"), false);

		// Application startup
		if (game.user.isGM) {
			scsApp.generateColors();
		} else {
			scsApp.phases.colors = game.settings.get(scsApp.ID, "colors");
		}
		scsApp.addPhases();
		scsApp.hideFromPlayers();
		scsApp.manageDisplay(html);
		scsApp.combatManager();
		api.changeVisibility(!!game.combat || !game.settings.get(scsApp.ID, "hideNoCombat"));

		// Pin zone is the "jiggle area" in which the app will be locked to a pinned position if dropped. pinZone stores whether or not we're currently in that area.
		let pinZone = false;

		// Have to override this because of the non-standard drag handle, and also to manage the pin lock zone and animation effects.
		drag._onDragMouseMove = function _newOnDragMouseMove(event) {
			event.preventDefault();

			const playerApp = document.getElementById("players");
			const playerAppPos = playerApp.getBoundingClientRect();

			// Limit dragging to 60 updates per second.
			const now = Date.now();
			if (now - this._moveTime < 1000 / 60) return;
			this._moveTime = now;

			// Follow the mouse.
			let conditionalOffset = 0;
			if (game.settings.get(scsApp.ID, "pinned")) {
				conditionalOffset = 20;
			}

			if (document.getElementById("scs-pin-lock")) scsApp.unPinApp(this);

			this.app.setPosition({
				left: this.position.left + (event.clientX - this._initial.x),
				top: this.position.top + (event.clientY - this._initial.y - conditionalOffset),
			});

			let rowsHeight = 46 * Math.min(Math.ceil(scsApp.phases.count / 3), 3); // Height of all rows of phases

			// Defining a region above the PlayerList that will trigger the jiggle.
			let playerAppUpperBound = playerAppPos.top - scsApp.pinOffset - rowsHeight;
			let playerAppLowerBound = playerAppPos.top + scsApp.pinOffset;

			if (event.clientX < 215 && event.clientY > playerAppUpperBound && event.clientY < playerAppLowerBound) {
				document.querySelector("#scsApp").style.animation = "jiggle 0.2s infinite";
				pinZone = true;
			} else {
				document.querySelector("#scsApp").style.animation = "";
				pinZone = false;
			}
		};

		drag._onDragMouseUp = async function _newOnDragMouseUp(event) {
			event.preventDefault();

			window.removeEventListener(...this.handlers.dragMove);
			window.removeEventListener(...this.handlers.dragUp);

			const playerApp = document.getElementById("players");
			const playerAppPos = playerApp.getBoundingClientRect();
			let rowsHeight = 46 * Math.min(Math.ceil(scsApp.phases.count / 3), 3); // Height of all rows of phases
			let myOffset = playerAppPos.height + scsApp.pinOffset + rowsHeight + 12;

			// If the mouseup happens inside the Pin zone, pin the app.
			if (pinZone) {
				scsApp.pinApp(true);
				await game.settings.set(scsApp.ID, "pinned", true);
				this.app.setPosition({
					left: 15,
					top: window.innerHeight - myOffset,
				});
			} else {
				let windowPos = document.querySelector("#scsApp").getBoundingClientRect();
				let newPos = { top: windowPos.top, left: windowPos.left };
				await game.settings.set(scsApp.ID, "position", newPos);
				await game.settings.set(scsApp.ID, "pinned", false);
			}

			// Kill the jiggle animation on mouseUp.
			document.querySelector("#scsApp").style.animation = "";
		};

		// Re-pin the app when the player list is re-rendered
		Hooks.on("renderPlayerList", () => {
			if (game.settings.get(scsApp.ID, "pinned")) {
				scsApp.unPinApp(false);
				scsApp.pinApp();
			}
		});
	}

	/** Pin the app above the Players list **/
	static async pinApp() {
		// Only do this if a pin lock isn"t already in place.
		if (!document.querySelectorAll("#scs-pin-lock").length) {
			const playerApp = document.getElementById("players");
			const playerAppPos = playerApp.getBoundingClientRect();
			let rowsHeight = 46 * Math.min(Math.ceil(scsApp.phases.count / 3), 3); // Height of all rows of phases
			let myOffset = playerAppPos.height + scsApp.pinOffset + rowsHeight + 12;

			// Add the pin lock to the DOM
			const pinLock = document.createElement("style");
			pinLock.id = "scs-pin-lock";
			pinLock.innerHTML = `#scsApp {
                top: calc(100vh - ${myOffset}px)!important;
                left: 15px!important;
            }`;
			document.body.append(pinLock);

			await game.settings.set(scsApp.ID, "pinned", true);
		}
	}

	/** Un-pin the app */
	static unPinApp(app) {
		// Set initial position to what is being enforced by pin lock if triggered by app
		if (app) {
			let pinnedApp = document.getElementById("scsApp");
			app.position.left = pinnedApp.offsetLeft;
			app.position.top = pinnedApp.offsetTop;
		}

		// Remove the style tag that's pinning the window.
		document.querySelector("#scs-pin-lock").remove();
	}

	/** Initialize phase names */
	static initPhaseNames() {
		// Get phase names
		let settings = game.settings.get(scsApp.ID, "names");

		// Check if the user reset to default phase names
		if (settings === "undefined") {
			// If so, set to the default phases
			scsApp.phases.names = game.i18n.localize("scs.settings.phaseNames.defaults");
		} else {
			// If not, parse and use the first 50 new values from settings
			scsApp.phases.names = settings
				.split(",")
				.slice(0, 50)
				.map(val => val.trim());
		}
	}

	/**
	 * All allowed hues
	 * @param {number[]} bases - Base hues to check against
	 * @returns {number[]} Hues that validate certain rules from a given base hue
	 */
	static allAllowedHues = bases =>
		Array.from(new Array(360)) // Hue is in degrees
			.map((_, i) => i + 1) // Insert numbers into the Array from 1 to 300
			.filter(
				(
					h // Filter the Array
				) =>
					bases.deepFlatten().every(
						(
							base // Check if every base is compatible with this hue
						) =>
							Math.abs(base - h) > 15 && // Must be sufficiently different than the base
							Math.abs(base - h) <= 150 && // Must not be too different
							(h < 50 || h > 150) // Must not be green
					)
			);

	/** Last hue generated */
	static lastHue;

	/**
	 * Generate a hue that is compatible with given base hues
	 * @param {number[]} bases - Base hues to check against
	 * @return {number} A hue that is compatible with the given base hues
	 */
	static generateHue(...bases) {
		const validHues = scsApp.allAllowedHues(bases);
		const index = Math.floor(Math.random() * validHues.length);
		return (
			validHues[index] ??
			ui.notifications.error(
				`${scsApp.ID} | ${game.i18n.localize("scs.notifications.colorGeneration.tryingAgain")}`
			)
		);
	}

	/** Generate random beautiful color gradients */
	static generateColors() {
		// A hue for each phase
		scsApp.phases.colors.push(
			...scsApp.phases.names.reduce((accumulator, _name, i) => {
				let hueTop, hueBottom;

				// Try and assign existing hues for this phase
				try {
					hueTop = game.settings.get(scsApp.ID, "colors").map(([top]) => top)[i];
					hueBottom = game.settings.get(scsApp.ID, "colors").map(([, bottom]) => bottom)[i];
					if (!hueTop || !hueBottom) throw "no hues stored";
				} catch (err) {
					// If not, generate new hues

					// Generate the top hue from a random base
					hueTop = scsApp.generateHue(accumulator.at(-1) ?? Math.floor(Math.random() * 360));

					// Generate the bottom hue with the top hue as the base
					hueBottom = scsApp.generateHue(hueTop);
				}

				// Push the hues to temp storage
				accumulator.push([hueTop, hueBottom]);
				return accumulator;
			}, [])
		);

		// Update settings for storage
		game.settings.set(scsApp.ID, "colors", scsApp.phases.colors);
	}

	/** Adds the phases to the Application */
	static addPhases() {
		// Remove any existing buttons
		document.querySelectorAll("#scsApp .phase").forEach(button => {
			button.remove();
		});

		// Create "buttons" for phases
		scsApp.phases.names.forEach(name => {
			let phaseButton = document.createElement("div");
			document.querySelector("#scsApp .container").append(phaseButton);
			phaseButton.classList.add("phase");
			phaseButton.innerText = name;
		});

		// Add colors
		scsApp.phases.colors.forEach((color, i) => {
			document.querySelector("#scsApp .gradients").innerHTML += `
            .scs .phase:nth-child(${i + 1}) {
                background-image: linear-gradient(
                hsl(${color[0]} 80% 40%),
                hsl(${color[1]} 80% 40%)
                );
            }`;
		});
	}

	/** Hide buttons for players and re-adjust app size */
	static hideFromPlayers() {
		if (!game.user.isGM) {
			document.querySelectorAll("#scsApp .arrows > *.fas").forEach(arrow => {
				arrow.style.display = "none";
			});
		}
	}

	/** Manage phase display
	 * @param {HTMLElement} html The app's html
	 */
	static manageDisplay(html) {
		// Update on Render
		Hooks.on("renderscsApp", () => {
			scsApp.pullValues();
			scsApp.updateApp();
		});

		// Update for players
		Hooks.on("updateSetting", setting => {
			if (setting.data.key === "scs.currentPhase" && !game.user.isGM) {
				scsApp.pullValues();
				scsApp.updateApp();
			}
		});

		// Reset when a new combat is created
		Hooks.on("createCombat", () => {
			scsApp.pullValues();
			scsApp.updateApp();
		});

		// If GM, execute one of the functions below this, depending on the button clicked
		if (game.user.isGM) {
			html.querySelector(".lastRound").addEventListener("click", () => game.combat?.previousRound());
			html.querySelector(".lastPhase").addEventListener("click", () => api.changePhase(-1));
			html.querySelector(".nextPhase").addEventListener("click", () => api.changePhase(1));
			html.querySelector(".nextRound").addEventListener("click", () => game.combat?.nextRound());
			html.querySelectorAll(".phase").forEach((button, i) =>
				button.addEventListener("click", () => api.changePhase(i + 1 - scsApp.currentPhase))
			);
		}
	}

	/** Pull current phase and cycle */
	static pullValues() {
		scsApp.currentPhase = game.settings.get(scsApp.ID, "currentPhase"); // counts the current phase
		scsApp.currentCycle = game.settings.get(scsApp.ID, "currentCycle"); // counts the current cycle

		// Make sure that the phase is within bounds
		// This is needed because the total amount of phases might be changed by the user
		if (scsApp.currentPhase < 0) {
			scsApp.currentPhase = 1;
		} else if (scsApp.currentPhase > scsApp.phases.count + 1) {
			scsApp.currentPhase = scsApp.phases.count;
		}
	}

	/** Updates the app to display the correct state */
	static updateApp() {
		const buttons = document.querySelectorAll("#scsApp .phase"); // gets an array of the three buttons

		// Save new values if GM
		if (game.user.isGM) {
			game.settings.set(scsApp.ID, "currentPhase", scsApp.currentPhase);
			game.settings.set(scsApp.ID, "currentCycle", scsApp.currentCycle);
		}
		if (buttons[scsApp.currentPhase - 1]) {
			// Update the appearance of the buttons depending on the user's settings
			if (!game.settings.get(scsApp.ID, "alternateActive")) {
				// Active is darker
				buttons.forEach(current => {
					current.classList.remove("active");
				});
				buttons[scsApp.currentPhase - 1].classList.add("active");
			} else {
				// Active is lighter
				buttons.forEach(current => {
					current.classList.add("active");
				});
				buttons[scsApp.currentPhase - 1].classList.remove("active");
			}

			// Scroll the active phase button into view
			let scrollTarget = document.querySelector("#scsApp .phase.active");
			if (
				scrollTarget.parentNode.scrollTop + document.querySelector("#scsApp .container").clientHeight <
				scrollTarget.offsetTop
			) {
				scrollTarget.parentNode.scrollTop = scrollTarget.offsetTop - 4;
			} else if (scrollTarget.offsetTop < scrollTarget.parentNode.scrollTop) {
				scrollTarget.parentNode.scrollTop = scrollTarget.offsetTop - 4;
			}
		}
		// Update the Round number
		document.querySelector("#scsApp .round").innerHTML = [
			game.i18n.localize("COMBAT.Round"),
			game.combat?.round,
		].join(" ");
	}

	/** Manage combat */
	static async combatManager() {
		// Start a combat as soon as a Combatant is created
		Hooks.on("createCombatant", async () => {
			if (game.combat) {
				// If a combat exists, start it
				game.combat.startCombat();
			} else if (game.user.isGM) {
				// Other wise, create a new one if GM
				const combat = await getDocumentClass("Combat").implementation.create({
					scene: game.scenes.active?.id,
				});
				await combat?.activate({ render: false });
				ui.combat.initialize({ combat });
			}

			// If enabled, make the App visible
			if (game.settings.get(scsApp.ID, "hideNoCombat")) api.changeVisibility(true);
		});

		// Stop combatant if there aren't any combatants left
		Hooks.on("deleteCombatant", combatant => {
			const combat = combatant.parent;

			// Initialize and reset this pseudo-flag
			if (!game.combat.data.flags.hasOwnProperty("scs")) game.combat.data.flags.scs = {};
			combat.data.flags.scs.deleting = undefined;

			// Wait for after this Hook has been fired
			setTimeout(() => {
				// If all of the Combatants have been removed, this combat is being deleted for the first time, and this is the GM
				if (combat?.turns.length === 0 && !combat.data.flags.scs.deleting && game.user.isGM) {
					// Mark this combat as in deletion
					combat.data.flags.scs.deleting = true;

					// End the combat
					combat.endCombat();

					// If enabled, hide the App
					if (game.settings.get(scsApp.ID, "hideNoCombat")) api.changeVisibility(false);
				}
			}, 0);
		});

		// If enabled, make the App visible when a combat is created
		Hooks.on("createCombat", () => {
			if (game.settings.get(scsApp.ID, "hideNoCombat")) api.changeVisibility(true);
		});

		// If enabled, hide the App when all of the combats have been deleted
		Hooks.on("deleteCombat", () => {
			if (!game.combat && game.settings.get(scsApp.ID, "hideNoCombat")) api.changeVisibility(false);
		});

		Hooks.on("updateCombat", async (_document, change) => {
			scsApp.pullValues();
			scsApp.updateApp();

			// End combat if round is zero
			if (change.round === 0) game.combat?.endCombat();
		});

		// Create a new combat if there isn't one and start it
		if (!game.combat && game.settings.get(scsApp.ID, "hideTracker")) {
			const combat = await getDocumentClass("Combat").implementation.create({
				scene: canvas.scene?.id,
			});
			await combat?.activate({ render: false });
			ui.combat.initialize({ combat });
			game.combat?.startCombat();
		}
	}

	/** Lock users to only certain actions depending on the phase */
	static actionLocking() {
		// Check if enabled
		if (game.settings.get(scsApp.ID, "actionLocking")) {
			if (!["dnd5e"].includes(game.system.id)) {
				ui.notifications.notify(
					`${scsApp.ID} | ${game.i18n.format("scs.notifications.actionLocking.none", {
						title: game.system.data.title,
						context: "system",
					})}`
				);
			} else {
				// Register wrapper for Item Roll for Action Locking
				libWrapper.register(scsApp.ID, "CONFIG.Item.documentClass.prototype.roll", function (wrapped, ...args) {
					let thisPhase = scsApp.phases.names[scsApp.currentPhase - 1];
					// Don't change anything if this is not a known phase and notify user
					if (
						![
							game.i18n.localize("scs.settings.phaseNames.defaults.move"),
							game.i18n.localize("scs.settings.phaseNames.defaults.attacks"),
							game.i18n.localize("scs.settings.phaseNames.defaults.magic"),
						].includes(thisPhase)
					) {
						if (game.user.isGM)
							ui.notifications.notify(
								`${scsApp.ID} | ${game.i18n.format("scs.notifications.actionLocking.none", {
									title: thisPhase,
									context: "phase",
								})}`
							);
						return wrapped(...args);
					}

					// Manage action locking
					if (
						thisPhase === game.i18n.localize("scs.settings.phaseNames.defaults.move") &&
						(this.data.type === "spell" || this.hasAttack)
					) {
						// If it is currently the move phase and this is a spell or an attack, alert user and do nothing
						ui.notifications.error(
							`${scsApp.ID} | ${game.i18n.localize("scs.notifications.actionLocking.move")}`
						);
						return;
					} else if (
						thisPhase === game.i18n.localize("scs.settings.phaseNames.defaults.attacks") &&
						!this.hasAttack
					) {
						// If it is currently the attack phase and this is not an attack, alert user and do nothing
						ui.notifications.error(
							`${scsApp.ID} | ${game.i18n.localize("scs.notifications.actionLocking.attacks")}`
						);
						return;
					} else if (
						thisPhase === game.i18n.localize("scs.settings.phaseNames.defaults.magic") &&
						(this.data.type !== "spell" || this.hasAttack)
					) {
						// If it is currently the spells phase and this is not a spell or this has an attack, alert user and do nothing
						ui.notifications.error(
							`${scsApp.ID} | ${game.i18n.localize("scs.notifications.actionLocking.magic")}`
						);
						return;
					} else {
						// If not one of the cases above, allow action
						return wrapped(...args);
					}
				});
			}
		}
	}
}
